python/mpi.cpp

```diff
@@ -0,0 +1,99 @@
+#include <sstream>
+#include <string>
+
+#include <arbor/version.hpp>
+
+#include <pybind11/pybind11.h>
+
+#ifdef ARB_MPI_ENABLED
+#include <arbor/communication/mpi_error.hpp>
+
+#include <mpi.h>
+
+#include "mpi.hpp"
+
+#ifdef ARB_WITH_MPI4PY
+#include <mpi4py/mpi4py.h>
+#endif
+
+namespace pyarb {
+
+#ifdef ARB_WITH_MPI4PY
+
+mpi_comm_shim comm_from_mpi4py(pybind11::object& o) {
+    import_mpi4py();
+
+    // If object is already a mpi4py communicator, return
+    if (PyObject_TypeCheck(o.ptr(), &PyMPIComm_Type)) {
+        return mpi_comm_shim(*PyMPIComm_Get(o.ptr()));
+    }
+    throw arb::mpi_error(MPI_ERR_OTHER, "The argument is not an mpi4py communicator");
+}
+
+#endif
+
+// Some helper functions for initializing and finalizing MPI.
+// Arbor requires at least MPI_THREAD_SERIALIZED, because the communication task
+// that uses MPI can be run on any thread, and there will never be more than one
+// concurrent communication task.
+
```

Comment:
From Sam on Slack (to have it in one place): 
Had a chat with @bcumming about this this morning.

My position (and I think I convinced Ben) is that ultimately, we should either demand that the python user code handle all the MPI environment and just provide an interface for getting a communicator for the context  (the 'pure' approach), or if we're going to provide an easy interface, that we handle everything — initialization, clean up, and so on.

Exposing `MPI_Initialize` and `MPI_Finalize` is convenient from the point of view of writing the wrappers, but IMO is just pushing unnecessary complexity onto the user — the user isn't going to be able to do anything else useful with the MPI context we provide unless we start wrapping whole chunks of the MPI interface, which I think we all agreed should be avoided.

Concretely, we should keep the MPI_Initialize wrapper and friends for getting the PR through, but then we should change the interface so that one can initialize the arbor context with e.g. `mpi=auto`; the python library would then run the initialization if required, and set up a MPI_Finalize on clean-up.

Whether MPI is running or not, and which rank the python user code sees, can be exposed via the existing functionality that sits on `arb::execution_context`.